reference/language-concepts/namespaces.markdown

```diff
@@ -6,104 +6,76 @@ sorting: 100
+[%CFEngine_include_example(namespace_methods-usebundle.cf)%]
 
-```cf3
-    files:
-       "/file"
-          create => "true",
-           perms => name1:settings;
-```
+## Bodies
 
-If you don't make any namespace declarations, you'll be in the
-`default` namespace.  Bundles, bodies, classes, and variables from the
-`default` namespace can be accessed like any other:
+Bodies are assumed to be within the same namespace as the promiser. To use a body from another namespace the namespace must be specified by prefixing the body name with the namespace followed by a colon (`:`).
 
-```cf3
-    files:
-      "/file"
-         create => "true",
-          perms => default:settings;
-```
+A common mistake is forgetting to specify `default:` when using bodies from the standard library which resides in the `default` namespace.
 
-If you use the standard library from your own namespace, remember to
-specify this `default:` prefix.
+[%CFEngine_include_example(namespace_bodies.cf)%]
 
-To access classes, variables, or meta-data in bundles in a different namespace, use the
-colon as a namespace prefix:
+## Variables
 
-    $(namespace:bundle.variable)
-    $(namespace:bundle_meta.variable)
+Variables (except for Special Variables) are assumed to be within the same scope
+as the promiser but can also be referenced fully qualified with the namespace.
 
-**Note** that this means that if you are in a namespace that's not `default`, you *must* qualify classes from `default` fully:
+[%CFEngine_include_example(namespace_variable_references.cf)%]
 
-    default:myclass::
-    "do something" if => "default:myotherclass";
+[Special variables][Special Variables] are always accessible without a namespace
+  prefix. For example, `this`, `mon`, `sys`, and `const` fall in this category.
 
-### Namespacing of classes and variables created in policy
+[%CFEngine_include_example(namespace_special_var_exception.cf)%]
 
-In policy, you can't create classes outside your own namespace.  So
-the following, for example, will create the class `mynamespace:done`
-if it runs in the namespace `mynamespace`.
+**Notes:**
 
-```cf3
-    files:
-      "/file"
-         create => "true",
-         action => if_repaired("done");
-```
+- The [`variables` Augments key][Augments#variables] defines variables in the
+  `main` bundle of the `data` namespace by default but supports seeding variable
+  values in any specified namespace.
 
-Similarly, variables you create in a namespaced bundle have to be
-prefixed like `mynamespace:mybundle.myvar` from outside your
-namespace, but can use `mybundle.myvar` inside the namespace and
-`myvar` inside `mybundle`.
+## Classes
 
-As a workaround, you could have a helper bundle in another namespace
-to create classes and variables as needed.
+Promises can only define classes within the current namespace. Classes are
+understood to refer to classes in the current namespace if a namespace is not
```

Comment:
by "current namespace" here do you mean "default" or a special magical namespace that exists for the "current bundle or body"?

So without any body file control specifying a namespace 

```
bundle foo
{
  classes:
    "bar";
}
```

Would create `default:bar` or `foo.bar`? I know that vars and classes are a bit special because they are like static public global things which accessible "everywhere".

Reply:
> by "current namespace" here do you mean "default" or a special magical namespace that exists for the "current bundle or body"?

I mean current as in what this.namespace is at current point in policy.

> So without any body file control specifying a namespace 
> 
> ```
> bundle foo

Missing bundle type

> {
>   classes:
>     "bar";
> }
> ```
> 
> Would create `default:bar` 

Yes.

> or `foo.bar`? 

No, classees have a scope as in lifetime or visibility horizon. That's either bundle or namespace. A bundle scope class only lives or can be seen from that bundle or be undles it calls (if inheritance is explicitly requested).

A single class never has a dot in the name  the presence of a dot indicates a class expression (comparison of multiple classes).

foo.bar indicates a variable, but you said classes:.

> I know that vars and classes are a bit special because they are like static public global things which accessible "everywhere".

Reply:
Ok. so 

```
bundle agent foo
{
  vars:
    "bar" string => "yeah";
  classes:
    "baz";
}
```

Would 
- create `default:foo.bar` available everywhere and always
- create `baz` class but only visible in the foo bundle or bundles that foo calls with `methods` promises?

Reply:
> Ok. so
> 
> ```
> bundle agent foo
> {
>   vars:
>     "bar" string => "yeah";
>   classes:
>     "baz";
> }
> ```
> 
> Would
> 
> * create `default:foo.bar` available everywhere and always
Correct.
Furthermore, if your inside that bundle (that is to say, `$(this.namespace) == default` AND `$(this.bundle) == foo`)  you could refer to that variable value in one of three ways; `$(bar)`, `$(foo.bar)`, or `$(default:foo.bar)`. If you are in another bundle within the default namespace (that is to say, `$(this.namespace) == default` AND `$(this.bundle) != foo`) you could refer to the variable value in one of two ways; `$(foo.bar)`, or `$(default:foo.bar)`. And, if you are in some other namespace (that is to say, `$(this.namespace) != default`) you can refer to that variable value in only one way; `$(default:foo.bar)`

> * create `baz` class but only visible in the foo bundle or bundles that foo calls with `methods` promises?
Almost Correct.

The `baz` class would be visible inside the foo bundle  (that is to say, `$(this.namespace) == default` AND `$(this.bundle) == foo`) and inside any bundle that you call from that bundle with a `methods` promise which has `inherit => "true"`.

Furthermore, you could refer to that class from within the bundle as either simply `baz` or fully qualified as `default:baz`.

Reply:
Yes. All that is excellent explanation. :)